### PR TITLE
Training improvements

### PR DESCRIPTION
**Pull request by mparrett**
# Summary

Playing with a few things you might be interested in.

Key changes:

- Went with a smaller size for the model after running into overfitting.
- Added early stopping based on eval loss
- Reverted training hyperparams to defaults, validated with custom penalty disabled
- Incorporated custom penalty in training and generation
- Added a few more prompts to the generation along with different generation params

### CODE REVIEW

**Comment by mparrett** [comment]
Mainly wanted to show you https://github.com/mparrett/kilter_brain_gen/pull/1/commits/af783183ce3058a226cf1e2315401ae8a178ea6a

Sorry for the double PR noise!

**Comment by mparrett** [comment]
```
python3 src/clm/clm.py
```

```
0.70	train_epoch_5.839416058394161.0.2.a20.d15: p1452r15p1270r13p1387r14p1391r14p1234r13p1233r13p1233r13p1233r13p1232r13p1163r12p1387r14p1459r15p1283r13p1340r13p1352r13p1216r13p1387r14
```

image

**Comment by rparrett** [comment]
Love the sampling during training. I did wacky thing with `watchdog` over in the name model and this seems way nicer.

> Went with a smaller size for the model after running into overfitting.

Do you reckon that this was producing "good routes" mostly by virtue of the overfitting?

I trained / ran locally and see this:
```
0.90	k=50, p=0.95	prompt='a20d20'	clm-model.0.0.a20.d20,p1353r13p1534r15p1532r15p1457r15p1384r14p1455r15p1389r14p1450r15p1506r15p1543r15p1453r15p1458r15
1.10	k=50, p=0.95	prompt='a20'	clm-model.3.0.a20.unk,p1169r12p1142r12p1084r15p1149r12p1202r13p1230r13p1306r13p1387r14p1379r14p1383r14p1504r15p1534r15
1.10	k=50, p=0.95	prompt='a40'	clm-model.2.2.a40.unk,p1096r15p1097r15p1230r13p1302r13p1284r13p1391r14p1387r14p1485r15p1464r15p1534r15p1385r14p1513r15p1390r14
1.20	k=50, p=0.95	prompt='a40d15'	clm-model.1.1.a40.d15,p1391r14p1462r15p1532r15p1514r15p1459r15p1525r15p1456r15p1464r15p1458r15p1460r15p1462r15
1.30	temp=0.7	prompt='a40d15'	clm-model.1.0.a40.d15,p1391r14p1385r14p1379r14p1456r15p1387r14p1461r15p1542r15p1506r15p1457r15p1353r13
1.30	temp=0.7	prompt='a20d20'	clm-model.0.0.a20.d20,p1353r13p1470r15p1462r15p1459r15p1392r14p1458r15p1461r15p1456r15
1.30	k=50, p=0.95	prompt='a20d20'	clm-model.0.3.a20.d20,p1504r15p1525r15p1383r14p1458r15p1464r15p1454r15p1506r15p1379r14p1454r15p1505r15
1.30	num_beams=1	prompt='a40'	clm-model.2.4.a40.unk,p1525r15p1531r15p1459r15p1390r14p1383r14p1390r14p1496r15p1495r15p1451r15p1353r13p1532r15p1506r15p1390r14
```

The lowest-penalty route there has no start holds, which is a bit sad.

**Comment by rparrett** [comment]
Also, I unfortunately did a *bunch* of refactoring in the main branch after opening #2

**Comment by rparrett** [comment]
I disabled the custom loss by swapping in `Trainer` for `CustomTrainer`, and the penalty scores of the resulting routes seem comparable.

```
0.80	k=50, p=0.95	prompt='a40'	clm-model.2.3.a40.unk,p1458r15p1496r15p1389r14p1506r15p1454r15p1391r14p1489r15p1469r15p1485r15p1452r15p1542r15p1547r15p1385r14p1513r15p1455r15p1531r15
1.00	k=50, p=0.95	prompt='a40d15'	clm-model.1.3.a40.d15,p1382r14p1457r15p1451r15p1385r14p1485r15p1531r15p1461r15p1521r15p1459r15p1504r15p1457r15p1457r15p1464r15p1470r15
1.00	k=50, p=0.95	prompt='a40'	clm-model.2.4.a40.unk,p1149r12p1198r13p1267r13p1284r13p1283r13p1287r13p1456r15p1387r14p1391r14p1458r15p1388r14p1487r15
1.10	num_beams=1	prompt='a40d15'	clm-model.1.0.a40.d15,p1454r15p1387r14p1464r15p1461r15p1470r15p1386r14p1478r15p1461r15p1504r15p1532r15p1521r15p1452r15
1.20	num_beams=1	prompt='a20d20'	clm-model.0.4.a20.d20,p1485r15p1503r15p1478r15p1521r15p1469r15p1506r15p1387r14p1455r15p1385r14p1462r15
1.20	k=50, p=0.95	prompt='a20d20'	clm-model.0.1.a20.d20,p1459r15p1383r14p1459r15p1497r15p1479r15p1513r15p1461r15p1455r15p1454r15p1457r15p1457r15p1531r15
1.20	k=50, p=0.95	prompt='a20d20'	clm-model.0.4.a20.d20,p1455r15p1460r15p1478r15p1514r15p1487r15p1391r14p1531r15p1496r15p1478r15p1504r15p1389r14p1383r14p1388r14
1.20	num_beams=1	prompt='a40d15'	clm-model.1.1.a40.d15,p1470r15p1462r15p1534r15p1455r15p1516r15p1478r15p1540r15p1504r15p1452r15p1470r15p1391r14
```

**Comment by mparrett** [comment]
> Do you reckon that this was producing "good routes" mostly by virtue of the overfitting?

Possibly. Can't say with a whole lot of certainty because I hadn't been looking very critically at the generated routes. This might be one of the things that at one point got me interested in doing similarity search... since I seem to recall what appeared to be at least partial memorization of "good" routes from the training data.

**Comment by mparrett** [comment]
> Also, I unfortunately did a _bunch_ of refactoring in the main branch after opening #2

I'm happy to graft anything back in to main at some point. I was in a bit of a hurry last night and didn't catch this before doing a bunch of things. :)

**Comment by mparrett** [comment]
> The lowest-penalty route there has no start holds, which is a bit sad.

That is a bit sad and concerning. My initial reaction is to make the penalty for that higher. But I'm guessing it's not that simple. FWIW, last night I was tinkering mostly with sanity checking training prior to re-introducing the custom penalty. So I'm not sure it's really in a great place yet.

**Comment by rparrett** [comment]
I'll continue tinkering as well.

One thing I noticed while playing around is that I would like to also be logging the un-alpha-multiplied `custom_penalty`.

**Comment by rparrett** [comment]
I'm seeing some weirdness with the new logging -- `custom_penality` and friends values seemingly stop updating in tensorboard after a few initial logs.

I think I may have resolved this by doing:
```python
def log(self, logs):
    if not self.model.training:
        phase = "eval_"
    else:
        phase = ""

    logs[f"{phase}original_loss"] = self.original_loss
    logs[f"{phase}custom_penalty"] = self.custom_penalty
    logs[f"{phase}custom_penalty_div_a"] = self.custom_penalty_div_a

    super().log(logs)
```
but I'm not actually sure why this was happening

**Comment by mparrett** [comment]
Closing this PR in favor of a new similar PR built on top of current `main` branch
